browser/playlist/test/playlist_download_request_manager_browsertests.cc

```diff
@@ -371,3 +438,54 @@ IN_PROC_BROWSER_TEST_F(PlaylistDownloadRequestManagerBrowserTest,
            .duration = ""},
       });
 }
+
+IN_PROC_BROWSER_TEST_F(PlaylistDownloadRequestManagerBrowserTest,
+                       DynamicallyAddedMedia) {
+  testing::NiceMock<playlist::MockObserver> observer;
+  auto* playlist_service =
+      playlist::PlaylistServiceFactory::GetForBrowserContext(
+          browser()->profile());
+  ASSERT_TRUE(playlist_service);
+  playlist_service->AddObserver(observer.GetRemote());
+
+  const auto& url = LoadHTML(
+      R"html(
+        <html>
+        <meta property="og:image" content="/img.jpg">
+        <body>
+          <video>
+          </video>
+        </body>
+        <script>
+          // Attach video tag after a few seconds.
+          document.addEventListener('DOMContentLoaded', function() {
+            setTimeout(function() {
+              let videoElement = document.createElement('video');
+              videoElement.src = 'test1.mp4';
+              document.body.appendChild(videoElement);
+            }, 5000);
+          });
+        </script>
+        </html>
+      )html",
+      GURL("https://youtube.com")  // For now, detecting dynamically added media
+                                   // files is done only for Background web
+                                   // contents. So we should pretend to be
+                                   // Youtube.
+  );
+
+  // At the first execution, we don't get any media files.
+  // And after that, we should find dynamically added media files and
+  // notify observers.
+  using testing::_;
+  base::RunLoop run_loop;
+  ON_CALL(observer, OnMediaFilesUpdated(_, _)).WillByDefault([&]() {
+    run_loop.Quit();
+  });
+  EXPECT_CALL(observer, OnMediaFilesUpdated(url, _));
+  playlist_service->FindMediaFilesFromActiveTab(base::BindLambdaForTesting(
```

Comment:
1. can we do `EXPECT_CALL().WillOnce(... { run_loop_Quit(); })` instead of `ON_CALL + EXPECT_CALL`?
2. pls replace `base::BindLambdaForTesting` with `MockCallback` (`base/test/mock_callback.h`). It's a bad practice to ASSERT/EXPECT in a callback that might not be called at all.

Reply:
Thanks a lot! Didn't know those.